### PR TITLE
feat(datadog_archives sink): Add `gcp_cloud_storage` support

### LINKED ISSUE #8931: Add `gcp_cloud_storage` support to the `datadog_archives` sink

### PR DESCRIPTION
**Pull request by vladimir-dd**
Closes #8931 

1. Refactors GCS sink into a stream sink and extracts common logic.
2. Adds GCS suport to `datadog_archives` sink.

### CODE REVIEW

**Comment by vladimir-dd** [comment]
@tobz could you take a look please at the final version after merging your changes?

File: src/sinks/datadog_archives.rs [hunk 1]
```diff
@@ -118,8 +141,21 @@ enum ConfigError {
...
+/// batch timeout - 15min
+const DEFAULT_BATCH_SETTINGS: BatchSettings<()> = {
+    BatchSettings::const_default()
+        .bytes(100_000_000)
+        .timeout(900)
+};
+
 impl DatadogArchivesSinkConfig {
-    fn new(&self, cx: SinkContext) -> crate::Result<(super::VectorSink, super::Healthcheck)> {
+    async fn bulid_sink(
```
Comment (tobz) [inline src/sinks/datadog_archives.rs:155]:
```suggestion
    async fn build_sink(

File: src/sinks/gcs_common/service.rs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+#[derive(Debug)]
+pub struct GcsResponse {
+    pub inner: http::Response<Body>,
+}
+
+impl AsRef<EventStatus> for GcsResponse {
+    fn as_ref(&self) -> &EventStatus {
+        &EventStatus::Delivered
+    }
+}
```
Comment (tobz) [inline src/sinks/gcs_common/service.rs:77]: Once you merge in `master` and resolve the conflicts, the need for this specific impl will go away, but you'll need to impl `DriverReponse` instead, which is used by `Driver`.

File: src/sinks/mod.rs [hunk 3]
```diff
@@ -35,6 +35,8 @@ pub mod elasticsearch;
 pub mod file;
 #[cfg(feature = "sinks-gcp")]
 pub mod gcp;
+#[cfg(any(feature = "sinks-gcp", feature = "sinks-datadog_archives"))]
```
Comment (tobz) [inline src/sinks/mod.rs:38]: You can simply have it look for `sinks-gcp` since you include that feature flag via `sinks-datadog_archives`.

**Review by tobz — APPROVED** [review]
Looks good to me overall.

You still have some changes that will have to be made once you merge `master` to resolve the conflicts, but other than that, seems pretty straightforward to me overall.

File: src/sinks/mod.rs (hunk 3 above)
Comment (vladimir-dd) [reply src/sinks/mod.rs:38]: indeed, updated

File: src/sinks/datadog_archives.rs (hunk 1 above)
Comment (vladimir-dd) [reply src/sinks/datadog_archives.rs:155]: thanks, updated

File: src/sinks/gcs_common/service.rs (hunk 2 above)
Comment (vladimir-dd) [reply src/sinks/gcs_common/service.rs]: Updated.

**Review by tobz — APPROVED** [review]
Changes after merging in master look good. 👍🏻

**Pull request merged**